### PR TITLE
KAFKA-7103: Use bulkloading for RocksDBSegmentedBytesStore during init

### PR DESCRIPTION
**Pull request by Ishiihara**
This PR uses bulk loading for recovering RocksDBWindowStore, same as RocksDBStore. Will perform unit test and integration test. 

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by Ishiihara** [comment]
@guozhangwang @mjsax @abbccdda

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (abbccdda) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: typo: toggleForBulkLoading

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (Ishiihara) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java:142]: @guozhangwang I am not sure why flush was needed in the original implementation.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (shawnsnguyen) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: toggleForBulkLoading*

**Comment by Ishiihara** [comment]
@guozhangwang @mjsax Can you help take a look when you have time? Thanks!

**Comment by mjsax** [comment]
@Ishiihara Can you please create a JIRA for this. By default, all changes should have a JIRA. Thx.

Please also update the PR title accordingly. Thx.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: nit: add `final`

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: nit: add `final`
(also within the loop)

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: Can this be private?

I am wondering, if we could refactor and merge this with `RocksDBBatchingRestoreCallback` ?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: nit: add `final`

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: nit: add `final`

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: nit: add `final`

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: Can this be simplified using Java 8 methods?

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java]: nit: add `final`

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java]: nit: add `final`

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java]: nit: add `final`

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java]: nit: add `final`

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java]: nit: add `final`

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java]: Why do we need this three lines? Seems to be unrelated to the test? Or do we need this to get an actual segment? If yes, maybe we should create multiple segments?

**Review by mjsax — COMMENTED** [review]
Thank for the PR. Mostly nits.

**Comment by mjsax** [comment]
\cc @bbejeck @vvcephei

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: I agree. It would be good to have one callback for restoring a RocksDB store.

**Review by bbejeck — COMMENTED** [review]
Overall lgtm, but I think it's worth looking into refactoring to have one restore callback.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (Ishiihara) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: I prefer to leave it separate. If we merge it, we need to pass in segments and rocksDB to the constructor, which will make things more complex. Also, the RocksDBSegmentBytes store and RocksDB store are actually different.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (Ishiihara) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: Yes. This is good suggestion.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java
Comment (Ishiihara) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java]: We need to put a key to the RocksDBSegmentedByteStore to actually get a segment. Yeah. we can definitely create multiple segments.

**Review by Ishiihara — COMMENTED** [review]
@mjsax Would you mind to take another look?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (Ishiihara) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: I revert the change as it seems that Java does not provide a method of getOrElseUpdate as scala did.

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: Can't we use `writeBatchMap.computeIfAbsent(segment, s -> new WriteBatch());` ?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java:142]: Me neither. @guozhangwang can you confirm that it's ok to remove this?

**Review by mjsax — COMMENTED** [review]
Thanks for the update.

**Comment by Ishiihara** [comment]
@mjsax I updated the title to include the JIRA number. 
@bbejeck For this PR, I would like to focus on the correctness and testing. I filed **https://issues.apache.org/jira/browse/KAFKA-7105** to track the refactoring.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (Ishiihara) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: @mjsax Good suggestion!

**Comment by Ishiihara** [comment]
@mjsax @bbejeck @guozhangwang Would you please take another look? Thanks!
cc @abbccdda

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java
Comment (bbejeck) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java:357]: Nice test!

**Review by bbejeck — APPROVED** [review]
@Ishiihara thanks for the patch, LGTM

Looking more into the details of how a segment store works vs a non-segment store it makes sense to create another Jira to merge the two restore callbacks.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: Could this be package private instead of protected to me more restrictive?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java:142]: Checked on the source code history, I think it is piggy backed from the old manner, in which we do restoration along with the registration of the store (before 0.11.0) whereas now we only register the store but delay restoration in main loop. So I think we can safely remove this.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: nit: add `final`

nit: add space `entry : writeBatchMap.entrySet()` (I thought this was a checkstyle rule? Wondering if my memory is wrong because build passed)

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: as above

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java]: as above

File: streams/src/main/java/org/apache/kafka/streams/state/internals/Segment.java
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/Segment.java]: nit: remove `this` (code style: we try to avoid `this` is possible)

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java:142]: Thanks for confirming @guozhangwang !

**Review by mjsax — COMMENTED** [review]
Just some more nits. Overall LGTM. Will merge after addressed.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (Ishiihara) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java:142]: Thanks for confirming!

**Comment by Ishiihara** [comment]
@mjsax @bbejeck Thanks for the review. I addressed the review comments. PTAL.

**Comment by Ishiihara** [comment]
@mjsax @guozhangwang I discovered an edge case that that might not enable enable the bulk loading for segments that are not at the current on the machine.

**Comment by guozhangwang** [comment]
Could you elaborate a bit more?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (Ishiihara) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java:202]: @guozhangwang In case of a newly created segment, we need to call bulk loading explicitly since the segments.AllSegments() only return the segments that are already exists on the machine.

Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java:202]: Great catch! How did you discover this? It is covered in the unit tests (ie, the changes to applied?)

Comment (Ishiihara) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java:202]: @mjsax I enhanced the test to include this case. I discovered this case with some internal applications at Pinterest.

**Review by mjsax — APPROVED** [review]
LGTM.

@guozhangwang Can we merge this or do you have any further comments?

**Comment by guozhangwang** [comment]
@mjsax EDIT: I have one question about the special handling logic above. Otherwise LGTM.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java
Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java:202]: One question: it seems the `bulkLoadSegments` set will only keep increasing in size since we only initialize it once in `bulkLoadSegments = new HashSet<>(segments.allSegments());` and then do this ad-hoc addition here, when segments are dropped they will not be removed from `bulkLoadSegments` as well, right?

I'm thinking, if we could just call `final Set<Segment> bulkLoadSegments = new HashSet<>(segments.allSegments());` here every time, since 1) if the store does not exist yet, the `getOrCreateSegmentIfLive` will call openDB that makes the open flag for the newly created store, and 2) if the store does exist already, then `toggleForBulkLoading -> segment.toggleDbForBulkLoading` will make sure the store is open here.

Comment (Ishiihara) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java:202]: @guozhangwang Thanks for the comment. I addressed the comment to create a new set each time we create a new segment.

**Comment by Ishiihara** [comment]
@guozhangwang @mjsax I addressed Guozhang's comment. Could you take another look?

**Comment by guozhangwang** [comment]
retest this please

**Comment by guozhangwang** [comment]
@mjsax could you give it another look?

**Comment by Ishiihara** [comment]
@guozhangwang @mjsax Not sure why Jenkins is keep failing. The failure seems not related to my change. Also, please take another look when you have time. Thanks!

**Comment by guozhangwang** [comment]
retest this please

**Comment by Ishiihara** [comment]
retest this please

**Comment by guozhangwang** [comment]
retest this please

**Comment by guozhangwang** [comment]
@Ishiihara The jenkins failures are due to

```
/home/jenkins/jenkins-slave/workspace/kafka-pr-jdk10-scala2.12/streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java:212: error: unreported exception RocksDBException; must be caught or declared to be thrown
18:24:17                     batch.remove(record.key);
18:24:17                                 ^
18:24:17 /home/jenkins/jenkins-slave/workspace/kafka-pr-jdk10-scala2.12/streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java:214: error: unreported exception RocksDBException; must be caught or declared to be thrown
18:24:17                     batch.put(record.key, record.value);
```

**Comment by guozhangwang** [comment]
Made another pass, except the compilation error it lgtm. @Ishiihara please let me know if you can fix the compilation errors, after that I'll go and merge it.

**Comment by Ishiihara** [comment]
@guozhangwang Updated the PR to catch RocksDB exception. PTAL.

**Comment by guozhangwang** [comment]
retest this please

**Comment by guozhangwang** [comment]
Triggered system test https://jenkins.confluent.io/job/system-test-kafka-branch-builder/1854/ before merging.

**Review by mjsax — APPROVED** [review]
LGTM.

Thanks for the patch @Ishiihara!

**Comment by guozhangwang** [comment]
System test passed.

**Pull request merged**
